core/include/gnuradio-4.0/Block.hpp

```diff
@@ -1058,49 +1058,90 @@ struct Block : protected std::tuple<Arguments...> {
                 // space_available_on_output_ports is true by construction of samplesToProcess
             }
 
-        } else { // end of kIsSourceBlock
-            ports_status.in_samples  = std::min(ports_status.in_samples, std::min(requested_work, ports_status.in_samples_to_eos_tag));
-            ports_status.out_samples = ports_status.in_samples;
-            if (ports_status.in_samples_to_eos_tag == 0 || ports_status.in_samples_to_eos_tag < ports_status.in_min_samples) {
-                state = LifeCycleState::STOPPED;
-                updateInputAndOutputTags();
+        } else {                                                                         // end of kIsSourceBlock
+            ports_status.in_samples = std::min(ports_status.in_samples, requested_work); // clamp input to scheduler work constraint
+#ifdef _DEBUG
+            fmt::println("block {} - mark1 - in {} -> out {}", name, ports_status.in_samples, ports_status.out_samples);
+#endif
+            if constexpr (!Resampling::kEnabled) { // clamp input until (excluding) next tag unless the read-position is on the tag
+                // N.B. minimum size '1' because we could have two tags on adjacent samples.
+                ports_status.in_samples  = ports_status.nSamplesToNextTag == 0 ? std::min(ports_status.in_samples, std::max(ports_status.nSamplesToNextTagAfter, 1UZ))
+                                                                               : std::min(ports_status.in_samples, ports_status.nSamplesToNextTag);
+                ports_status.out_samples = ports_status.in_samples;
```

Comment:
It looks like this line can be removed because you update `ports_status.in_samples` directly after `if constexpr`, then `ports_status.out_samples` must be updated after as well.